File: python/examples/08_advanced_theory.py
#!/usr/bin/env python3
"""
Advanced Music Theory Analysis

This example demonstrates advanced music theory functionality:
- Scale degree analysis in various contexts
- Voice leading analysis and validation
- Harmonic function classification
- Tritone substitution validation
- Advanced chord substitutions and transformations
- Cross-scale modulation analysis

Based on the Composer specification: chord-theory-core.spec
"""

import composer


def demonstrate_scale_degree_analysis() -> None:
    """Show comprehensive scale degree analysis."""
    print("=== Scale Degree Analysis ===")

    major_scale = composer.ScaleFingerprint.major()
    minor_scale = composer.ScaleFingerprint.minor()

    # Test chords for scale degree analysis
    test_chords = [
        composer.Chord(1, 5),  # I major
        composer.Chord(2, 7),  # ii7
        composer.Chord(5, 7),  # V7
        composer.Chord(6, 5),  # vi
        composer.Chord(4, 9),  # IV9
    ]

    print("Scale degree analysis in major scale:")
    for chord in test_chords:
        print(f"  {chord}:")

        try:
            # Get stable scale degrees (absolute)
            stable_degrees = composer.get_stable_scale_degrees(chord, major_scale)
            print(f"    Stable scale degrees: {stable_degrees}")

            # Get relative scale degrees
            relative_degrees = composer.get_relative_scale_degrees(chord)
            print(f"    Relative degrees: {relative_degrees}")

            # Show degree numbers and accidentals separately if available
            if hasattr(relative_degrees, "sd_numbers") and hasattr(
                relative_degrees, "sd_accs"
            ):
                print(f"    Degree numbers: {relative_degrees.sd_numbers}")
                print(f"    Accidentals: {relative_degrees.sd_accs}")

        except AttributeError:
            # Fallback analysis
            print(f"    Root: {chord.root}, Type: {chord.chord_type}")
            print(f"    Inversion: {chord.inversion}")

            # Manual scale degree calculation
            base_degrees = {
                5: [1, 3, 5],  # Triad
                7: [1, 3, 5, 7],  # Seventh
                9: [1, 3, 5, 7, 9],  # Ninth
                11: [1, 3, 5, 7, 9, 11],  # Eleventh
                13: [1, 3, 5, 7, 9, 11, 13],  # Thirteenth
            }

            if chord.chord_type in base_degrees:
                chord_degrees = base_degrees[chord.chord_type]
                # Transpose by root
                transposed = [
                    (degree + chord.root - 2) % 7 + 1 for degree in chord_degrees
                ]
                print(f"    Calculated degrees: {transposed}")

        print()

    print("Comparing scale degree analysis across different scales:")
    test_chord = composer.Chord(5, 7)  # V7 chord
    scales = [
        ("Major", major_scale),
        ("Natural Minor", minor_scale),
        ("Dorian", composer.ScaleFingerprint.dorian()),
        ("Mixolydian", composer.ScaleFingerprint.mixolydian()),
    ]

    for scale_name, scale in scales:
        print(f"  {test_chord} in {scale_name}:")

        try:
            stable_degrees = composer.get_stable_scale_degrees(test_chord, scale)
            print(f"    Scale degrees: {stable_degrees}")
        except AttributeError:
            print(f"    Scale: {scale}")

        print()


def demonstrate_harmonic_functions() -> None:
    """Show harmonic function classification."""
    print("=== Harmonic Function Analysis ===")

    # Define harmonic functions
    harmonic_functions = {
        "Tonic": [1, 6, 3],  # I, vi, iii
        "Predominant": [4, 2],  # IV, ii
        "Dominant": [5, 7],  # V, vii°
    }

    print("Harmonic function classification:")
    for function_name, scale_degrees in harmonic_functions.items():
        print(f"  {function_name} function:")

        for degree in scale_degrees:
            # Create different chord types on each degree
            chord_types = [5, 7]  # Triad and seventh

            for chord_type in chord_types:
                chord = composer.Chord(degree, chord_type)
                print(f"    {chord} (scale degree {degree})")

                try:
                    # Analyze harmonic function
                    function = composer.analyze_harmonic_function(chord, "major")
                    print(f"      Function: {function}")

                except AttributeError:
                    # Manual function classification
                    if degree in harmonic_functions["Tonic"]:
                        function = "Tonic"
                    elif degree in harmonic_functions["Predominant"]:
                        function = "Predominant"
                    elif degree in harmonic_functions["Dominant"]:
                        function = "Dominant"
                    else:
                        function = "Chromatic"

                    print(f"      Function: {function}")

                # Show complexity
                complexity = composer.get_chord_complexity(chord)
                print(f"      Complexity: {complexity:.2f}")

        print()


def demonstrate_voice_leading() -> None:
    """Show voice leading analysis and validation."""
    print("=== Voice Leading Analysis ===")

    # Common chord progressions with different voice leading qualities
    progressions = [
        {
            "name": "Good voice leading (I-vi-IV-V)",
            "chords": [
                composer.Chord(1, 5),  # I
                composer.Chord(6, 5),  # vi
                composer.Chord(4, 5),  # IV
                composer.Chord(5, 5),  # V
            ],
        },
        {
            "name": "Parallel motion (I-II-III)",
            "chords": [
                composer.Chord(1, 5),  # I
                composer.Chord(2, 5),  # II (parallel major chords)
                composer.Chord(3, 5),  # III
            ],
        },
        {
            "name": "Large leaps (I-♭VI-I)",
            "chords": [
                composer.Chord(1, 5),  # I
                composer.Chord(6, 5, borrowed="minor"),  # ♭VI
                composer.Chord(1, 5),  # I
            ],
        },
    ]

    print("Voice leading analysis:")
    for progression in progressions:
        print(f"  {progression['name']}:")
        chords = progression["chords"]

        print(f"    Progression: {[str(c) for c in chords]}")

        try:
            # Analyze voice leading quality
            quality = composer.analyze_voice_leading(chords)
            print(f"    Voice leading quality: {quality:.3f}")

            # Check for specific voice leading issues
            parallel_fifths = composer.check_parallel_fifths(chords)
            parallel_octaves = composer.check_parallel_octaves(chords)
            large_leaps = composer.check_large_leaps(chords)

            print(f"    Parallel fifths: {parallel_fifths}")
            print(f"    Parallel octaves: {parallel_octaves}")
            print(f"    Large leaps: {large_leaps}")

        except AttributeError:
            # Manual voice leading analysis
            total_complexity = sum(composer.get_chord_complexity(c) for c in chords)
            avg_complexity = total_complexity / len(chords)

            # Simple heuristics
            print(f"    Average complexity: {avg_complexity:.2f}")

            # Check for chromatic movement
            roots = [c.root for c in chords]
            chromatic_steps = sum(
                1 for i in range(1, len(roots)) if abs(roots[i] - roots[i - 1]) == 1
            )
            print(f"    Chromatic steps: {chromatic_steps}")

            # Estimate voice leading quality
            quality_estimate = max(0, 1 - (avg_complexity - 2) / 8)
            print(f"    Estimated quality: {quality_estimate:.3f}")

        print()


def demonstrate_tritone_substitution() -> None:
    """Show tritone substitution analysis and validation."""
    print("=== Tritone Substitution Analysis ===")

    # Test chords for tritone substitution
    test_chords = [
        composer.Chord(5, 7),  # V7 (should be valid)
        composer.Chord(5, 7, applied=5),  # V7/V (should be valid)
        composer.Chord(5, 7, applied=6),  # V7/vi (should be valid)
        composer.Chord(1, 5),  # I (should not be valid)
        composer.Chord(2, 7),  # ii7 (should not be valid)
    ]

    print("Tritone substitution validation:")
    for chord in test_chords:
        print(f"  {chord}:")

        try:
            # Check if tritone substitution is valid
            is_valid = composer.is_valid_tri_sub(chord, "major")
            print(f"    Valid for tritone substitution: {is_valid}")

            if is_valid:
                # Calculate the tritone substitute
                substitute = composer.get_tritone_substitute(chord)
                print(f"    Tritone substitute: {substitute}")

        except AttributeError:
            # Manual tritone sub validation
            is_dominant = chord.chord_type == 7 and (
                chord.root == 5 or chord.applied != 0
            )

            print(f"    Is dominant seventh: {is_dominant}")
            print(f"    Applied chord: {chord.applied != 0}")

            if is_dominant:
                # Calculate manual substitute (tritone away)
                sub_root = (chord.root + 3) % 7 + 1  # Tritone = 3 scale degrees
                print(f"    Would substitute with root: {sub_root}")

        print()


def demonstrate_chord_substitutions() -> None:
    """Show various chord substitution techniques."""
    print("=== Chord Substitution Techniques ===")

    original_progression = [
        composer.Chord(1, 5),  # I
        composer.Chord(6, 5),  # vi
        composer.Chord(4, 5),  # IV
        composer.Chord(5, 5),  # V
    ]

    print(f"Original progression: {[str(c) for c in original_progression]}")
    print()

    # Different substitution techniques
    substitution_techniques = [
        {
            "name": "Relative minor/major substitution",
            "substitutions": {
                0: composer.Chord(6, 5),  # I -> vi
                1: composer.Chord(1, 5),  # vi -> I
            },
        },
        {
            "name": "Modal interchange (borrowed chords)",
            "substitutions": {
                2: composer.Chord(4, 5, borrowed="minor"),  # IV -> iv
            },
        },
        {
            "name": "Secondary dominants",
            "substitutions": {
                3: composer.Chord(5, 7, applied=6),  # V -> V7/vi
            },
        },
        {
            "name": "Extended harmonies",
            "substitutions": {
                0: composer.Chord(1, 7),  # I -> IM7
                3: composer.Chord(5, 9),  # V -> V9
            },
        },
    ]

    for technique in substitution_techniques:
        print(f"  {technique['name']}:")

        # Apply substitutions
        substituted = original_progression.copy()
        for index, substitute_chord in technique["substitutions"].items():
            original_chord = substituted[index]
            substituted[index] = substitute_chord

            print(f"    Position {index + 1}: {original_chord} -> {substitute_chord}")

            # Analyze the substitution
            try:
                # Check if chords are isotonal
                major_scale = composer.ScaleFingerprint.major()
                are_isotonal = composer.is_isotonal(
                    original_chord, substitute_chord, major_scale
                )
                print(f"      Isotonal: {are_isotonal}")

            except AttributeError:
                # Compare complexity
                orig_complexity = composer.get_chord_complexity(original_chord)
                sub_complexity = composer.get_chord_complexity(substitute_chord)

                complexity_change = f"{orig_complexity:.2f} -> {sub_complexity:.2f}"
                print(f"      Complexity change: {complexity_change}")

        print(f"    Result: {[str(c) for c in substituted]}")
        print()


def demonstrate_modulation_analysis() -> None:
    """Show analysis of modulation between keys/scales."""
    print("=== Modulation Analysis ===")

    # Common modulation progression: C major to G major
    modulation_progression = [
        composer.Chord(1, 5),  # I in C (C major)
        composer.Chord(6, 5),  # vi in C (A minor)
        composer.Chord(2, 7),  # ii7 in C (D minor 7) - becomes vi7 in G
        composer.Chord(5, 7),  # V7 in C (G7) - becomes I7 in G
        composer.Chord(1, 5),  # I in G (G major) - new key
    ]

    scales = [
        composer.ScaleFingerprint.major(),  # C major
        composer.ScaleFingerprint.major(),  # G major (same pattern, different tonic)
    ]

    print("Analyzing modulation from C major to G major:")
    print(f"Progression: {[str(c) for c in modulation_progression]}")
    print()

    # Analyze each chord in both key contexts
    for i, chord in enumerate(modulation_progression):
        print(f"  Chord {i + 1}: {chord}")

        # Analyze in C major context
        try:
            degrees_c = composer.get_stable_scale_degrees(chord, scales[0])
            print(f"    In C major: {degrees_c}")
        except AttributeError:
            print(f"    In C major: root {chord.root}")

        # Analyze in G major context (transpose analysis)
        try:
            # For G major, we need to transpose the analysis
            degrees_g = composer.get_stable_scale_degrees(chord, scales[1])
            print(f"    In G major: {degrees_g}")
        except AttributeError:
            # Manual transposition
            transposed_root = (chord.root + 4) % 7 + 1  # G is 5th degree of C
            print(f"    In G major: root {transposed_root}")

        # Determine pivot chord quality
        if i == 2 or i == 3:  # These are pivot chords
            print("    Pivot chord function")

        print()

    # Analyze modulation smoothness
    print("Modulation analysis:")
    total_complexity = sum(
        composer.get_chord_complexity(c) for c in modulation_progression
    )
    avg_complexity = total_complexity / len(modulation_progression)

    print(f"  Average complexity: {avg_complexity:.2f}")
    print(f"  Modulation smoothness: {'Smooth' if avg_complexity < 3 else 'Complex'}")
    print()


def demonstrate_cross_scale_analysis() -> None:
    """Show analysis of chords across different scale types."""
    print("=== Cross-Scale Analysis ===")

    test_chord = composer.Chord(5, 7)  # V7 chord

    # Analyze the same chord in different scale contexts
    scale_contexts = [
        ("Major", composer.ScaleFingerprint.major()),
        ("Natural Minor", composer.ScaleFingerprint.minor()),
        ("Dorian", composer.ScaleFingerprint.dorian()),
        ("Mixolydian", composer.ScaleFingerprint.mixolydian()),
        (
            "Harmonic Minor",
            composer.ScaleFingerprint(
                [
                    True,
                    False,
                    True,
                    True,
                    False,
                    True,
                    False,
                    True,
                    True,
                    False,
                    False,
                    True,
                ]
            ),
        ),
    ]

    print(f"Analyzing {test_chord} across different scales:")
    print()

    for scale_name, scale in scale_contexts:
        print(f"  In {scale_name} scale:")
        print(f"    Scale pattern: {scale.chromatic_notes}")

        try:
            stable_degrees = composer.get_stable_scale_degrees(test_chord, scale)
            print(f"    Scale degrees: {stable_degrees}")

            # Check if chord fits naturally in scale
            chord_notes = set(stable_degrees)
            scale_notes = set(
                str(i) for i, present in enumerate(scale.chromatic_notes, 1) if present
            )

            chord_in_scale = chord_notes.issubset(scale_notes)
            print(f"    Fits in scale: {chord_in_scale}")

        except AttributeError:
            # Manual analysis
            print(f"    Root: {test_chord.root}, Type: {test_chord.chord_type}")

            # Count how many chord tones are in the scale
            chord_tones = 4 if test_chord.chord_type == 7 else 3  # Rough estimate
            print(f"    Estimated chord tones: {chord_tones}")

        # Function in this scale context
        if test_chord.root == 5:
            if scale_name == "Major":
                function = "Dominant (V7)"
            elif scale_name == "Mixolydian":
                function = "Tonic (I7)"
            else:
                function = "Fifth degree seventh"
        else:
            function = f"Degree {test_chord.root} seventh"

        print(f"    Harmonic function: {function}")
        print()


def main() -> None:
    """Run all advanced theory demonstration functions."""
    print("Composer Library - Advanced Music Theory Examples")
    print("=" * 60)
    print()

    demonstrate_scale_degree_analysis()
    demonstrate_harmonic_functions()
    demonstrate_voice_leading()
    demonstrate_tritone_substitution()
    demonstrate_chord_substitutions()
    demonstrate_modulation_analysis()
    demonstrate_cross_scale_analysis()

    print("All advanced theory examples completed!")
    print()
    print("Note: Some advanced theory features may require additional Python")
    print("bindings to be fully functional. This example demonstrates the")
    print("intended usage patterns based on the specification.")


if __name__ == "__main__":
    main()

File: wasm/examples/typescript/08_advanced_theory.ts
#!/usr/bin/env ts-node
/**
 * Advanced Music Theory Analysis
 *
 * This example demonstrates advanced music theory functionality:
 * - Scale degree analysis in various contexts
 * - Voice leading analysis and validation
 * - Harmonic function classification
 * - Tritone substitution validation
 * - Advanced chord substitutions and transformations
 * - Cross-scale modulation analysis
 *
 * Based on the Composer specification: chord-theory-core.spec
 */

import * as composer from "../../composer_wasm";

interface ProgressionInfo {
  name: string;
  chords: composer.WasmChord[];
}

interface HarmonicFunction {
  name: string;
  scaleDegrees: number[];
}

/**
 * Initialize the WASM module
 */
async function initializeWasm(): Promise<void> {
  console.log(`Initializing Composer WASM v${composer.getVersion()}`);
}

/**
 * Show comprehensive scale degree analysis
 */
function demonstrateScaleDegreeAnalysis(): void {
  console.log("=== Scale Degree Analysis ===");

  const majorScale = composer.WasmScaleFingerprint.major();
  const minorScale = composer.WasmScaleFingerprint.minor();

  // Test chords for scale degree analysis
  const testChords = [
    { chord: new composer.WasmChord(1, 5), name: "I major" },
    { chord: new composer.WasmChord(2, 7), name: "ii7" },
    { chord: new composer.WasmChord(5, 7), name: "V7" },
    { chord: new composer.WasmChord(6, 5), name: "vi" },
    { chord: new composer.WasmChord(4, 9), name: "IV9" },
  ];

  console.log("Scale degree analysis in major scale:");
  testChords.forEach(({ chord, name }) => {
    console.log(`  ${name}:`);

    try {
      // Get stable scale degrees (absolute)
      const scaleDegrees = composer.getStableScaleDegrees(chord, majorScale);
      console.log(`    Stable scale degrees: [${scaleDegrees.join(", ")}]`);

      // Show chord properties
      console.log(`    Root: ${chord.root}, Type: ${chord.chordType}`);
      console.log(`    Inversion: ${chord.inversion}`);

      // Calculate complexity
      const complexity = composer.getChordComplexity(chord, "major");
      console.log(`    Complexity: ${complexity.toFixed(2)}`);
    } catch (error) {
      console.log(`    Analysis error: ${error}`);
    }

    console.log();
  });

  console.log("Comparing scale degree analysis across different scales:");
  const testChord = new composer.WasmChord(5, 7); // V7 chord
  const scales = [
    { name: "Major", scale: majorScale },
    { name: "Natural Minor", scale: minorScale },
  ];

  scales.forEach(({ name, scale }) => {
    console.log(`  ${testChord.toString()} in ${name}:`);

    try {
      const scaleDegrees = composer.getStableScaleDegrees(testChord, scale);
      console.log(`    Scale degrees: [${scaleDegrees.join(", ")}]`);
    } catch (error) {
      console.log(`    Analysis error: ${error}`);
    }

    console.log();
  });

  // Clean up objects
  testChords.forEach(({ chord }) => chord.free());
  testChord.free();
  majorScale.free();
  minorScale.free();
}

/**
 * Show harmonic function classification
 */
function demonstrateHarmonicFunctions(): void {
  console.log("=== Harmonic Function Analysis ===");

  // Define harmonic functions
  const harmonicFunctions: HarmonicFunction[] = [
    { name: "Tonic", scaleDegrees: [1, 6, 3] }, // I, vi, iii
    { name: "Predominant", scaleDegrees: [4, 2] }, // IV, ii
    { name: "Dominant", scaleDegrees: [5, 7] }, // V, vii°
  ];

  console.log("Harmonic function classification:");
  harmonicFunctions.forEach(({ name, scaleDegrees }) => {
    console.log(`  ${name} function:`);

    scaleDegrees.forEach((degree) => {
      // Create different chord types on each degree
      const chordTypes = [5, 7]; // Triad and seventh

      chordTypes.forEach((chordType) => {
        try {
          const chord = new composer.WasmChord(degree, chordType);
          console.log(`    ${chord.toString()} (scale degree ${degree})`);

          // Manual function classification
          let functionName: string;
          if (harmonicFunctions[0].scaleDegrees.includes(degree)) {
            functionName = "Tonic";
          } else if (harmonicFunctions[1].scaleDegrees.includes(degree)) {
            functionName = "Predominant";
          } else if (harmonicFunctions[2].scaleDegrees.includes(degree)) {
            functionName = "Dominant";
          } else {
            functionName = "Chromatic";
          }

          console.log(`      Function: ${functionName}`);

          // Show complexity
          const complexity = composer.getChordComplexity(chord, "major");
          console.log(`      Complexity: ${complexity.toFixed(2)}`);

          chord.free();
        } catch (error) {
          console.log(`    Error creating chord ${degree}/${chordType}: ${error}`);
        }
      });
    });

    console.log();
  });
}

/**
 * Show voice leading analysis and validation
 */
function demonstrateVoiceLeading(): void {
  console.log("=== Voice Leading Analysis ===");

  // Common chord progressions with different voice leading qualities
  const progressions: ProgressionInfo[] = [
    {
      name: "Good voice leading (I-vi-IV-V)",
      chords: [
        new composer.WasmChord(1, 5), // I
        new composer.WasmChord(6, 5), // vi
        new composer.WasmChord(4, 5), // IV
        new composer.WasmChord(5, 5), // V
      ],
    },
    {
      name: "Parallel motion (I-II-III)",
      chords: [
        new composer.WasmChord(1, 5), // I
        new composer.WasmChord(2, 5), // II (parallel major chords)
        new composer.WasmChord(3, 5), // III
      ],
    },
    {
      name: "Large leaps (I-♭VI-I)",
      chords: [
        new composer.WasmChord(1, 5), // I
        new composer.WasmChord(6, 5), // ♭VI (simulated)
        new composer.WasmChord(1, 5), // I
      ],
    },
  ];

  console.log("Voice leading analysis:");
  progressions.forEach(({ name, chords }) => {
    console.log(`  ${name}:`);
    console.log(`    Progression: ${chords.map((c) => c.toString())}`);

    // Manual voice leading analysis
    const totalComplexity = chords.reduce((sum, chord) => {
      try {
        return sum + composer.getChordComplexity(chord, "major");
      } catch (error) {
        return sum;
      }
    }, 0);
    const avgComplexity = totalComplexity / chords.length;

    console.log(`    Average complexity: ${avgComplexity.toFixed(2)}`);

    // Check for chromatic movement
    const roots = chords.map((c) => c.root);
    const chromaticSteps = roots.reduce((count, root, i) => {
      if (i > 0) {
        const interval = Math.abs(root - roots[i - 1]);
        return count + (interval === 1 ? 1 : 0);
      }
      return count;
    }, 0);
    console.log(`    Chromatic steps: ${chromaticSteps}`);

    // Estimate voice leading quality
    const qualityEstimate = Math.max(0, 1 - (avgComplexity - 2) / 8);
    console.log(`    Estimated quality: ${qualityEstimate.toFixed(3)}`);

    console.log();
  });

  // Clean up objects
  progressions.forEach(({ chords }) => {
    chords.forEach((chord) => chord.free());
  });
}

/**
 * Show tritone substitution analysis and validation
 */
function demonstrateTritoneSubstitution(): void {
  console.log("=== Tritone Substitution Analysis ===");

  // Test chords for tritone substitution
  const testChords = [
    { chord: new composer.WasmChord(5, 7), name: "V7", shouldBeValid: true },
    { chord: new composer.WasmChord(1, 5), name: "I", shouldBeValid: false },
    { chord: new composer.WasmChord(2, 7), name: "ii7", shouldBeValid: false },
    { chord: new composer.WasmChord(7, 7), name: "vii7", shouldBeValid: false },
  ];

  console.log("Tritone substitution validation:");
  testChords.forEach(({ chord, name, shouldBeValid }) => {
    console.log(`  ${name} (${chord.toString()}):`);

    // Manual tritone sub validation
    const isDominantSeventh = chord.chordType === 7 && chord.root === 5;

    console.log(`    Is dominant seventh: ${isDominantSeventh}`);
    console.log(`    Expected validity: ${shouldBeValid}`);
    console.log(`    Actual validity: ${isDominantSeventh}`);

    if (isDominantSeventh) {
      // Calculate manual substitute (tritone away)
      const subRoot = ((chord.root + 3) % 7) + 1; // Tritone = 3 scale degrees (simplified)
      console.log(`    Tritone substitute root: ${subRoot}`);

      try {
        const substitute = new composer.WasmChord(subRoot, 7);
        console.log(`    Tritone substitute: ${substitute.toString()}`);
        substitute.free();
      } catch (error) {
        console.log(`    Error creating substitute: ${error}`);
      }
    }

    console.log();
  });

  // Clean up objects
  testChords.forEach(({ chord }) => chord.free());
}

/**
 * Show various chord substitution techniques
 */
function demonstrateChordSubstitutions(): void {
  console.log("=== Chord Substitution Techniques ===");

  const originalProgression = [
    new composer.WasmChord(1, 5), // I
    new composer.WasmChord(6, 5), // vi
    new composer.WasmChord(4, 5), // IV
    new composer.WasmChord(5, 5), // V
  ];

  console.log(`Original progression: ${originalProgression.map((c) => c.toString())}`);
  console.log();

  // Different substitution techniques
  const substitutionTechniques = [
    {
      name: "Relative minor/major substitution",
      substitutions: [
        { position: 0, newChord: new composer.WasmChord(6, 5) }, // I -> vi
        { position: 1, newChord: new composer.WasmChord(1, 5) }, // vi -> I
      ],
    },
    {
      name: "Extended harmonies",
      substitutions: [
        { position: 0, newChord: new composer.WasmChord(1, 7) }, // I -> IM7
        { position: 3, newChord: new composer.WasmChord(5, 9) }, // V -> V9 (if supported)
      ],
    },
  ];

  substitutionTechniques.forEach((technique) => {
    console.log(`  ${technique.name}:`);

    // Apply substitutions
    const substituted = [...originalProgression];
    technique.substitutions.forEach(({ position, newChord }) => {
      const originalChord = substituted[position];
      console.log(
        `    Position ${position + 1}: ${originalChord.toString()} -> ${newChord.toString()}`
      );

      // Analyze the substitution
      try {
        const origComplexity = composer.getChordComplexity(originalChord, "major");
        const subComplexity = composer.getChordComplexity(newChord, "major");

        const complexityChange = `${origComplexity.toFixed(2)} -> ${subComplexity.toFixed(2)}`;
        console.log(`      Complexity change: ${complexityChange}`);

        // Simulate isotonal check using scale degrees
        const majorScale = composer.WasmScaleFingerprint.major();
        try {
          const origDegrees = composer.getStableScaleDegrees(originalChord, majorScale);
          const subDegrees = composer.getStableScaleDegrees(newChord, majorScale);
          const sharedDegrees = origDegrees.filter((d: string) => subDegrees.includes(d));
          const areIsotonal = sharedDegrees.length >= 2;
          console.log(`      Isotonal (simulated): ${areIsotonal}`);
        } catch (error) {
          console.log(`      Isotonal analysis failed: ${error}`);
        }
        majorScale.free();
      } catch (error) {
        console.log(`      Analysis error: ${error}`);
      }

      substituted[position] = newChord;
    });

    console.log(`    Result: ${substituted.map((c) => c.toString())}`);
    console.log();

    // Clean up substitution chords
    technique.substitutions.forEach(({ newChord }) => newChord.free());
  });

  // Clean up original progression
  originalProgression.forEach((chord) => chord.free());
}

/**
 * Show analysis of modulation between keys/scales
 */
function demonstrateModulationAnalysis(): void {
  console.log("=== Modulation Analysis ===");

  // Common modulation progression: C major to G major
  const modulationProgression = [
    new composer.WasmChord(1, 5), // I in C (C major)
    new composer.WasmChord(6, 5), // vi in C (A minor)
    new composer.WasmChord(2, 7), // ii7 in C (D minor 7) - becomes vi7 in G
    new composer.WasmChord(5, 7), // V7 in C (G7) - becomes I7 in G
    new composer.WasmChord(1, 5), // I in G (G major) - new key
  ];

  const majorScale = composer.WasmScaleFingerprint.major();

  console.log("Analyzing modulation from C major to G major:");
  console.log(`Progression: ${modulationProgression.map((c) => c.toString())}`);
  console.log();

  // Analyze each chord in both key contexts
  modulationProgression.forEach((chord, i) => {
    console.log(`  Chord ${i + 1}: ${chord.toString()}`);

    // Analyze in C major context
    try {
      const degrees = composer.getStableScaleDegrees(chord, majorScale);
      const roman = composer.getRomanNumeral(chord, majorScale);
      console.log(`    In C major: ${roman} (degrees: [${degrees.join(", ")}])`);
    } catch (error) {
      console.log(`    In C major: Analysis error - ${error}`);
    }

    // Manual analysis for G major context (simplified)
    // In a real implementation, this would require transposition
    const transposedRoot = ((chord.root + 4) % 7) + 1; // G is 5th degree of C
    console.log(`    In G major: estimated root degree ${transposedRoot}`);

    // Determine pivot chord quality
    if (i === 2 || i === 3) {
      // These are pivot chords
      console.log("    *** Pivot chord function ***");
    }

    console.log();
  });

  // Analyze modulation smoothness
  console.log("Modulation analysis:");
  const totalComplexity = modulationProgression.reduce((sum, chord) => {
    try {
      return sum + composer.getChordComplexity(chord, "major");
    } catch (error) {
      return sum;
    }
  }, 0);
  const avgComplexity = totalComplexity / modulationProgression.length;

  console.log(`  Average complexity: ${avgComplexity.toFixed(2)}`);
  console.log(`  Modulation smoothness: ${avgComplexity < 3 ? "Smooth" : "Complex"}`);
  console.log();

  // Clean up objects
  modulationProgression.forEach((chord) => chord.free());
  majorScale.free();
}

/**
 * Show analysis of chords across different scale types
 */
function demonstrateCrossScaleAnalysis(): void {
  console.log("=== Cross-Scale Analysis ===");

  const testChord = new composer.WasmChord(5, 7); // V7 chord

  // Analyze the same chord in different scale contexts
  const scaleContexts = [
    { name: "Major", scale: composer.WasmScaleFingerprint.major() },
    { name: "Natural Minor", scale: composer.WasmScaleFingerprint.minor() },
    { name: "Harmonic Minor", scale: composer.WasmScaleFingerprint.harmonicMinor() },
  ];

  console.log(`Analyzing ${testChord.toString()} across different scales:`);
  console.log();

  scaleContexts.forEach(({ name, scale }) => {
    console.log(`  In ${name} scale:`);
    console.log(`    Scale note count: ${scale.noteCount()}`);
    console.log(`    Is diatonic: ${scale.isDiatonic()}`);

    try {
      const scaleDegrees = composer.getStableScaleDegrees(testChord, scale);
      console.log(`    Scale degrees: [${scaleDegrees.join(", ")}]`);

      // Check if chord fits naturally in scale (simplified analysis)
      const chordInScale = scaleDegrees.length > 0;
      console.log(`    Fits in scale: ${chordInScale}`);
    } catch (error) {
      console.log(`    Analysis error: ${error}`);
    }

    // Function in this scale context
    let harmonicFunction: string;
    if (testChord.root === 5) {
      if (name === "Major") {
        harmonicFunction = "Dominant (V7)";
      } else if (name === "Mixolydian") {
        harmonicFunction = "Tonic (I7)";
      } else {
        harmonicFunction = "Fifth degree seventh";
      }
    } else {
      harmonicFunction = `Degree ${testChord.root} seventh`;
    }

    console.log(`    Harmonic function: ${harmonicFunction}`);
    console.log();
  });

  // Clean up objects
  testChord.free();
  scaleContexts.forEach(({ scale }) => scale.free());
}

/**
 * Run all advanced theory demonstration functions
 */
async function main(): Promise<void> {
  console.log("Composer Library - Advanced Music Theory Examples (TypeScript)");
  console.log("=".repeat(70));
  console.log();

  try {
    await initializeWasm();

    demonstrateScaleDegreeAnalysis();
    demonstrateHarmonicFunctions();
    demonstrateVoiceLeading();
    demonstrateTritoneSubstitution();
    demonstrateChordSubstitutions();
    demonstrateModulationAnalysis();
    demonstrateCrossScaleAnalysis();

    console.log("All advanced theory examples completed!");
    console.log();
    console.log("Note: Some advanced theory features may require additional WASM");
    console.log("bindings to be fully functional. This example demonstrates the");
    console.log("intended usage patterns based on current available functionality.");
  } catch (error) {
    console.error("Error running advanced theory examples:", error);
  }
}

// Run if executed directly
if (require.main === module) {
  main().catch(console.error);
}

// Export functions for potential reuse
export {
  demonstrateScaleDegreeAnalysis,
  demonstrateHarmonicFunctions,
  demonstrateVoiceLeading,
  demonstrateTritoneSubstitution,
  demonstrateChordSubstitutions,
  demonstrateModulationAnalysis,
  demonstrateCrossScaleAnalysis,
};
